File: 2/g.py
import sys

lines = sys.stdin.readlines()
curr_line = 0
first = True

while True:
    if curr_line > len(lines):
        break
    line= lines[curr_line]
    curr_line += 1

    if line == "0":
        break

    if first:
        first = False
    else:
        print()
    n, k =  [int(x) for x in line.split()]

    players = [[0,0] for _ in range(n)]


    for i in range(k*n*(n-1)//2):
        p_one, c_one, p_two, c_two = lines[curr_line].split()
        curr_line += 1

        p_one = int(p_one)
        p_two = int(p_two)

        winner = 0
        
        if c_two == c_one:
            winner = 0
        elif (c_one == "rock" and c_two == "scissors") or \
            (c_one == "paper" and c_two == "rock") or \
            (c_one == "scissors" and c_two == "paper"):
            winner = 1
        else:
            winner = 2

        if winner == 1:
            players[p_one-1][0] += 1
            players[p_two-1][1] += 1
        elif winner == 2:
            players[p_one-1][1] += 1
            players[p_two-1][0] += 1
    
    for player in players:
        try:
            avg = player[0]/(player[1]+player[0])
            print(f"{avg:.3f}")
        except:
            print('-')
        
        

File: 2/g.cpp
#include <algorithm>
#include <cmath>
#include <cstdint>
#include <cstdio>
#include <functional>
#include <iostream>
#include <iterator>
#include <map>
#include <set>
#include <sstream>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>
#include <iomanip>

#define sz(x) (int(x.size()))
#define fast_io()            \
  {                          \
    ios::sync_with_stdio(0); \
    cin.tie(NULL);           \
  }
#define mult_vec(name, T, n, m) \
  std::vector<std::vector<T>> name(n, std::vector<T>(m))
// #define DB

#ifdef DB
#define DEBUG(s) std::cout << s << '\n';
#else
#define DEBUG(s) ;
#endif

typedef long long ll;
typedef unsigned long long ull;
typedef std::vector<int> vi;
typedef std::vector<std::string> vs;
typedef std::vector<std::vector<std::string>> vvs;
typedef std::vector<char> vc;
typedef std::vector<std::vector<char>> vvc;
typedef std::vector<std::vector<std::string>> vvs;
typedef std::vector<bool> vb;
typedef std::vector<std::vector<bool>> vvb;
typedef std::vector<ll> vll;
typedef std::vector<ull> vull;
typedef std::pair<int, int> pii;
typedef std::pair<ll, ll> pll;
typedef std::pair<ull, ull> pull;
typedef std::vector<pii> vpii;
typedef std::vector<pll> vpll;
typedef std::vector<pull> vpull;
typedef std::vector<std::vector<int>> vvii;
typedef std::vector<std::vector<ll>> vvll;
typedef std::vector<std::vector<ull>> vvull;
typedef std::vector<std::vector<pii>> vvpii;
typedef std::vector<std::vector<pll>> vvpll;
typedef std::vector<std::vector<pull>> vvpull;

using namespace std;

int main() {
    fast_io();

    int n;
    bool first = true;
    while (cin >> n) {
        if (!n) break;

        if (first) first = false;
        else cout << '\n';

        int k;
        cin >> k;

        vpii players(n, {0,0});

        for (int i = 0; i < k*n*(n-1)/2; ++i) {
            int p1, p2;
            string c1, c2;
            cin >> p1 >> c1 >> p2 >> c2;

            int winner = 0;
            if (c1 == c2) continue;

            if ((c1 == "rock" && c2 == "scissors") || (c1 == "paper" && c2 == "rock") || (c1 == "scissors" && c2 == "paper"))
                winner = 1;
            else
                winner = 2;

            if (winner == 1) {
                players[p1-1].first++;
                players[p2-1].second++;
            } else {
                players[p1-1].second++;
                players[p2-1].first++;
            }
        }

        for (auto&[w, l] : players) {
            if (w + l == 0) cout << "-\n";
            else cout << fixed << setprecision(3) << ((double)w/(w+l)) << '\n';
        }
        
    }

    return 0;
}
